gen-ai/orchestrator-server/src/main/python/server/src/gen_ai_orchestrator/services/langchain/rag_chain.py

```diff
@@ -233,37 +237,72 @@ def create_rag_chain(query: RagQuery, vector_db_async_mode: Optional[bool] = Tru
     if query.compressor_setting:
         retriever = add_document_compressor(retriever, query.compressor_setting)
 
+    # Log progress and validate prompt template
+    logger.info('RAG chain - Validating LLM prompt template')
+    validate_prompt_template(query.question_answering_prompt)
+
     logger.debug('RAG chain - Document index name: %s', query.document_index_name)
-    logger.debug('RAG chain - Create a ConversationalRetrievalChain from LLM')
-
-    return ConversationalRetrievalChain.from_llm(
-        llm=llm_factory.get_language_model(),
-        retriever=retriever,
-        return_source_documents=True,
-        return_generated_question=True,
-        combine_docs_chain_kwargs={
-            'prompt': PromptTemplate(
-                template=llm_factory.setting.prompt,
-                input_variables=__find_input_variables(llm_factory.setting.prompt),
-            )
-        },
-    )
 
+    # Build LLM and prompt templates
+    llm = llm_factory.get_language_model()
+    rag_prompt = build_rag_prompt(query)
 
-def __find_input_variables(template):
-    """
-    Search for input variables on a given template
+    # Construct the RAG chain using the prompt and LLM
+    rag_chain = construct_rag_chain(llm, rag_prompt)
 
-    Args:
-        template: the template to search on
+    # Build the chat chain for question contextualization
+    chat_chain = build_chat_chain(llm)
+
+    # Function to contextualize the question based on chat history
+    contextualize_question_fn = partial(contextualize_question, chat_chain=chat_chain)
+
+    # Final RAG chain with retriever and source documents
+    rag_chain_with_source = contextualize_question_fn | RunnableParallel(
```

Comment:
A suggestion to be more explicit : 
```suggestion
    rag_chain_with_retriever = contextualize_question_fn | RunnableParallel(
```